std/protocols/permutation.asm

```diff
@@ -1,44 +1,149 @@
+        from_base(0)
+    };
+    let beta = Fp2::Fp2(beta1, beta2);
+    
+    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
+    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));
+    
+    // acc' = acc * lhs_folded / rhs_folded
+    let res = mul_ext(
+        eval_ext(mul_ext(acc, lhs_folded)),
+        inv_ext(eval_ext(rhs_folded))
+    );
+
+    match res {
+        Fp2::Fp2(a0_fe, a1_fe) => [a0_fe, a1_fe]
+    }
+};
+
+/// Adds constraints that enforce that lhs is a permutation of rhs
+/// Arguments:
+/// - acc: A phase-2 witness column to be used as the accumulator. If 2 are provided, computations
+///        are done on the F_{p^2} extension field.
+/// - permutation_constraint: The permutation constraint
+let permutation: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
+
+    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);
 
     let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = assert(modulus() > 2**100, || "This implementation assumes a large field");
+    let with_extension = match len(acc) {
+        1 => false,
+        2 => true,
+        _ => panic("Expected 1 or 2 accumulator columns!")
+    };
+
+    let _ = if !with_extension {
+        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
+    } else { [] };
+
+    // On the extension field, we'll need two field elements to represent the challenge.
+    // If we don't need an extension field, we can simply set the second component to 0,
+    // in which case the operations below effectively only operate on the first component.
+    let fp2_from_array = |arr| if with_extension { Fp2::Fp2(arr[0], arr[1]) } else { from_base(arr[0]) };
+    let acc_ext = fp2_from_array(acc);
+    let alpha = fp2_from_array([alpha1, alpha2]);
+    let beta = fp2_from_array([beta1, beta2]);
 
     // If the selector is 1, contribute a factor of `beta - compress_expression_array(lhs)` to accumulator.
-    // If the selector is 0, contribute a factor of 1 to the accumulator. 
-    let lhs_folded = lhs_selector * (beta - compress_expression_array(lhs) - 1) + 1;
-    let rhs_folded = rhs_selector * (beta - compress_expression_array(rhs) - 1) + 1;
+    // If the selector is 0, contribute a factor of 1 to the accumulator.
+    // Implemented as: folded = selector * (beta - compress_expression_array(values) - 1) + 1;
+    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
+    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));
+
+    let next_acc = if with_extension {
+        next_ext(acc_ext)
+    } else {
+        // The second component is 0, but the next operator is not defined on it...
```

Comment:
Why is that the case? Next applied on literals should just return the literal.

Reply:
I agree, but it doesn't!

When I do it I get:
```
thread 'main' panicked at /Users/georg/coding/powdr/pil-analyzer/src/condenser.rs:243:21:
Error reducing expression to constraint:
Expression: std::protocols::permutation::permutation([main.z], main.permutation_constraint)
Error: TypeError("Expected column for \"'\" operator, but got: 0")
```

I opened #1442.